### PR TITLE
Refactor and simplify string token handling and parse them correctly in all grammars

### PR DESCRIPTION
**Pull request by pablogsal**
Closes: #163 

Apart from simplifying things a bit, this PR makes sure that `string_token` *always* produces valid AST string nodes (`Constant`) with the appropriate string types. Then, in `concatenate_strings` we just add them together (if the types are correct) but we don't do any more parsing, which arguably is a better split of responsibilities between the functions.

### CODE REVIEW

**Comment by pablogsal** [comment]
With this PR:

```
venv ❯ make test         
python -c "from pegen import parse; t = parse.parse_file('data/cprog.txt'); exec(compile(t, '', 'exec'))"
Hello world
```

File: pegen/pegen.c [hunk 1]
```diff
@@ -381,26 +381,76 @@ number_token(Parser *p)
...
+
+    if (t == NULL) {
         return NULL;
-    // TODO: Creating an AST constant here and joining the strings afterwards is
-    // inefficient. We should just move around the char*.
-    PyObject *c = PyUnicode_FromStringAndSize(s, len);
-    if (!c)
+    }
+
+    char *the_str = PyBytes_AsString(t->bytes) ;
```
Comment (gvanrossum) [inline pegen/pegen.c]:
```suggestion
    char *the_str = PyBytes_AsString(t->bytes);
```

File: pegen/pegen.c [hunk 2]
```diff
@@ -381,26 +381,76 @@ number_token(Parser *p)
...
+    assert(s != NULL && fstr == NULL);
+    assert(bytesmode ? PyBytes_CheckExact(s) : PyUnicode_CheckExact(s));
+    final_str = s;
+
+    if (PyArena_AddPyObject(p->arena, final_str) < 0) {
+        goto error;
+    }
+
+    if (!bytesmode && kind_unicode) {
+        //TODO: Intern this string when we decide how we will
```
Comment (gvanrossum) [inline pegen/pegen.c]: Do you intentionally leave out the space between `//` and `TODO`?

File: pegen/pegen.c [hunk 3]
```diff
@@ -381,26 +381,76 @@ number_token(Parser *p)
...
+
+    char *the_str = PyBytes_AsString(t->bytes) ;
+    if (the_str == NULL) {
         return NULL;
-    if (PyArena_AddPyObject(p->arena, c) < 0) {
-        Py_DECREF(c);
+    }
+
+    PyObject *s;
+    PyObject *final_str = NULL;
```
Comment (gvanrossum) [inline pegen/pegen.c]: Why not move this declaration to the first use at line 434?

File: pegen/pegen.c [hunk 4]
```diff
@@ -381,26 +381,76 @@ number_token(Parser *p)
...
+    if (parsestr(p, the_str, &bytesmode, &this_rawmode, &s,
+                 &fstr, &fstrlen) != 0) {
         return NULL;
     }
-    return Constant(c, NULL, t->lineno, t->col_offset, t->end_lineno, t->end_col_offset, p->arena);
+
+    /* Check if it has a 'u' prefix */
+    int kind_unicode = 0;
+    if (the_str[0] == 'u') {
+        kind_unicode = 1;
```
Comment (gvanrossum) [inline pegen/pegen.c]: Maybe `assert(!bytesmode);` here? They're mutually exclusive.

File: pegen/pegen.c [hunk 5]
```diff
@@ -381,26 +381,76 @@ number_token(Parser *p)
...
 expr_ty
 string_token(Parser *p)
 {
     Token *t = expect_token(p, STRING);
-    if (t == NULL)
-        return NULL;
-    char *s = NULL;
-    Py_ssize_t len = 0;
-    if (PyBytes_AsStringAndSize(t->bytes, &s, &len) < 0)
+    PyObject *u_kind = NULL;
```
Comment (gvanrossum) [inline pegen/pegen.c]: You can move this to line 439. (I like declaring variables near the first point they're needed.)

File: pegen/pegen.c (hunk 2 above)
Comment (pablogsal) [reply pegen/pegen.c]: Nop, there should be a space there.

**Comment by pablogsal** [comment]
> Is there even a point in preserving the 'u' flag, given that we're only Parsing Python 3 and it's redundant?

Functionally not much, but helps with testing (more opportunities to test prefixes) and to mirror what the current AST in CPython produces (the `kind` field in string nodes). We could remove it in the future to simplify the code, indeed.

**Pull request merged**